File: core/contractsdk/go/example/short_content/short_content.go
package main

import (
	"errors"
	"github.com/xuperchain/xuperchain/core/contractsdk/go/code"
	"github.com/xuperchain/xuperchain/core/contractsdk/go/driver"
	"github.com/xuperchain/xuperchain/core/contractsdk/go/utils"
)

const (
	TOPIC_LENGTH_LIMIT   = 36
	TITLE_LENGTH_LIMIT   = 100
	CONTENT_LENGTH_LIMIT = 3000
	USER_BUCKET          = "USER"
)

var (
	ErrContentLengthTooLong = errors.New("the length of topic or title or content is more than limitation")
)

type shortContent struct {
}

func (sc *shortContent) Initialize(ctx code.Context) code.Response {
	return code.OK(nil)
}

func (sc *shortContent) StoreShortContent(ctx code.Context) code.Response {
	args := struct {
		UserId  []byte `json:"user_id",required:"true"`
		Title   []byte `json:"title",required:"true"`
		Topic   []byte `json:"topic",required:"true"`
		Content []byte `json;"content",required:"true"`
	}{}
	if err := utils.Validate(ctx.Args(), &args); err != nil {
		return code.Error(err)
	}
	userKey := utils.ConcatWithString(USER_BUCKET, "/", args.UserId, "/", args.Topic, "/", args.Title)
	if len(args.Topic) > TOPIC_LENGTH_LIMIT || len(args.Title) > TITLE_LENGTH_LIMIT ||
		len(args.Content) > CONTENT_LENGTH_LIMIT {
		//	TODO 注意下这里 len 的含义
		return code.Error(ErrContentLengthTooLong)
	}
	if err := ctx.PutObject(userKey, args.Content); err != nil {
		return code.Error(err)
	}
	return code.OK(nil)
}

func (sc *shortContent) QueryByUser(ctx code.Context) code.Response {
	args := struct {
		UserID []byte `json:"user_id",required:"true"`
	}{}
	if err := utils.Validate(ctx.Args(), &args); err != nil {
		return code.Error(err)
	}
	start := utils.ConcatWithString(USER_BUCKET, "/", args.UserID, "/")
	end := utils.ConcatWithString(start, "~")
	result := []byte{}
	iter := ctx.NewIterator(start, end)
	defer iter.Close()
	//iter 需要关闭
	for iter.Next() {
		result = append(result, utils.ConcatWithString(iter.Key(), "\n", iter.Value(), "\n")...)
	}
	return code.OK(result)
}
func (sc *shortContent) QueryByTitle(ctx code.Context) code.Response {
	args := struct {
		UserId []byte `json:"user_id",required:"true"`
		Topic  []byte `json:"user_id",required:"true"`
		Title  []byte `json:"title",required:"true"`
	}{}
	if err := utils.Validate(ctx.Args(), &args); err != nil {
		return code.Error(err)
	}
	value, err := ctx.GetObject(utils.ConcatWithString(USER_BUCKET, "/", args.UserId, "/", args.Topic, "/", args.Title))
	if err != nil {
		return code.Error(err)
	}
	return code.OK(value)
}

func (sc *shortContent) QueryByTopic(ctx code.Context) code.Response {
	args := struct {
		UserId []byte `json:"user_id",required:"true"`
		Topic  []byte `json:"user_id",required:"true"`
	}{}
	start := utils.ConcatWithString(USER_BUCKET, "/", args.UserId, "/", args.Topic, "/")
	end := utils.ConcatWithString(start, "~")
	iter := ctx.NewIterator(start, end)
	defer iter.Close()
	result := []byte{}

	for iter.Next() {
		result = append(result, utils.ConcatWithString(iter.Key(), "\n", iter.Value(), "\n")...)
	}
	return code.OK(result)
}

func main() {
	driver.Serve(new(shortContent))
}

File: core/contractsdk/java/example/short_content/src/main/java/com/baidu/xuper/ShortContent.java
package com.baidu.xuper;

import com.sun.tools.javac.util.ByteBuffer;

/**
 * Counter
 */
public class ShortContent implements Contract {
    final String userBucket = "USER";
    final int titleLengthLimit = 100;
    final int contentLengthLimit = 3000;

    @Override
    @ContractMethod
    public Response initialize(Context ctx) {
        return Response.ok("ok".getBytes());
    }

    @ContractMethod
    public Response storeShortContent(Context ctx) {
        ByteBuffer buf = new ByteBuffer();
        byte[]userId = ctx.args().get("user_id");
        byte[] title = ctx.args().get("title");
        byte[]topic =  ctx.args().get("topic");
        byte[]content = ctx.args().get("content");
//        TODO
        buf.appendBytes(userBucket.getBytes());
        buf.appendBytes("/".getBytes());
        buf.appendBytes(userId);
        buf.appendBytes("/".getBytes());
        buf.appendBytes(topic);
        buf.appendBytes("/".getBytes());
        buf.appendBytes(title);
        if( topic.length  > contentLengthLimit||title.length > titleLengthLimit || content.length > contentLengthLimit){
            return Response.error("The length of topic or title or content is more than limitation");
        }
        byte[] userKey = buf.elems;
//        error 处理呢？
        ctx.putObject(userKey,content);
        return Response.ok("ok".getBytes());

    }

    @ContractMethod
    public Response queryByUser(Context ctx){
        String userId = ctx.args().get("user_id").toString();
        ByteBuffer result = new ByteBuffer();
        String start = userBucket + "/"  + "user_id" + "/";
        String end = start + "~";
        ctx.newIterator(start.getBytes(),end.getBytes()).forEachRemaining(
                item ->{
                    result.appendBytes(item.getKey());
                    result.appendBytes("\n".getBytes());
                    result.appendBytes(item.getValue());
                    result.appendBytes("\n".getBytes());
                }
        );
        return Response.ok("ok".getBytes());
    }

    @ContractMethod
    public Response queryByTitle(Context ctx){
        String userId = ctx.args().get("user_id").toString();
        String  title = ctx.args().get("title").toString();
        String topic =  ctx.args().get("topic").toString();
//        String content = ctx.args().get("content").toString();
        String key = userBucket + "/" + userId + "/" + "/" + topic + "/" + title;
        return Response.ok(ctx.getObject(key.getBytes()));
    }

    @ContractMethod
    public Response queryByTopic(Context ctx){
        byte[] userId = ctx.args().get("user_id");
        byte[] topic = ctx.args().get("topic");
        ByteBuffer key = new ByteBuffer();
        key.appendBytes(userBucket.getBytes());
        key.appendBytes("/".getBytes());
        key.appendBytes(userId);
        key.appendBytes("/".getBytes());
        key.appendBytes(topic);
        key.appendBytes("/".getBytes());

        byte[] start = key.elems;
        key.appendBytes("~".getBytes());
        byte []end = key.elems;
        ByteBuffer result = new ByteBuffer();
        ctx.newIterator(start,end).forEachRemaining(
                item ->{
                    result.appendBytes(item.getKey());
                    result.appendBytes("\n".getBytes());
                    result.appendBytes(item.getValue());
                    result.appendBytes("\n".getBytes());
                }
        );
        return Response.ok(result.elems);
    }


    public static void main(String[] args) {
        Driver.serve(new ShortContent());
    }
}